src/jdk.attach/windows/classes/sun/tools/attach/VirtualMachineImpl.java

```diff
@@ -126,12 +126,12 @@ public SocketInputStreamImpl(long fd) {
         }
 
         @Override
-        protected int readImpl(long fd, byte[] bs, int off, int len) throws IOException {
+        protected int read(long fd, byte[] bs, int off, int len) throws IOException {
             return VirtualMachineImpl.readPipe(fd, bs, off, len);
```

Comment:
If you add read/close methods for Windows that then call readPipe/closePipe, then we don't need to subclass and override.

Reply:
Looks like Windows needs a long fd so the shared API takes long and we then cast to int for the Unix native methods. A little messy but fixable.

Reply:
> If you add read/close methods for Windows that then call readPipe/closePipe, then we don't need to subclass and override.

Sorry, I don't understand what this means. VirtualMachineImpl is platform-dependent while HotSpotVirtualMachine is shared, we always need to sub-class socket input stream in order to call platform-depend read/close.

Reply:
As long as all the platform-specific `VirtualMachineImpl` classes define a `read` and `close` method then we can simply have the shared socket stream class call those methods. Windows will need an extra level of indirection because its `read` method will have to call `readPipe`.

Reply:
> As long as all the platform-specific `VirtualMachineImpl` classes define a `read` and `close` method then we can simply have the shared socket stream class call those methods. Windows will need an extra level of indirection because its `read` method will have to call `readPipe`.

```
/path/src/jdk.attach/share/classes/sun/tools/attach/HotSpotVirtualMachine.java:444: error: incompatible types: possible lossy conversion from long to int
            return VirtualMachineImpl.read(fd, bs, off, len);
                                           ^
/path/src/jdk.attach/share/classes/sun/tools/attach/HotSpotVirtualMachine.java:449: error: incompatible types: possible lossy conversion from long to int
                VirtualMachineImpl.close(fd);
```
Unfortunately, this causes compilation error because we have `long fd` in shard socket stream and  `read(int fd...)/readPipe(long fd)` in VirtualMachineImpl

Reply:
That is unfortunate.